Meshcoord dataview daskidea

pp-mo opened the pull request:
## 🚀 Pull Request

### Description
A demonstration of how the MeshCoord points/bounds calculation could be done **with Dask**.

A new test module shows how we can construct an AuxCoord mimicking what we want from a MeshCoord,
and adds various tests that demonstrate desired behaviour.

One desirable thing that is _missing_, is that the dask operations used don't allow an efficient partial access : i.e. if you compute some portion of a result (even the points access, which is simpler), it _always_ fetches the whole data, before sub-indexing that.

We have found before that this kind of problem **_can_** often be fixed, with a different construction of the Dask calculation.
In any case, given the expected semi-random ordering of information relative to 'useful' data regions, we expect that effectively enabling this may require a more involved solution.
Having simply confirmed what it now does, I am leaving any further investigation to the future.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

stephenworsley commented on lib/iris/tests/unit/experimental/ugrid/test_meshcoord__dataview.py:
Range should vary with face index shape.

pp-mo replied:
Good spot.
Will do this

pp-mo commented:
It was noticed in discussion with @stephenworsley that 
  1. in `test_partial_access_points` I had as my "partial access" `mesh_coord.points[1:2]`.  Which of course fetches the whole data, and should really use `mesh_coord[1:2].points` or maybe `mesh_coord.lazy_points()[1:2].compute`.
 2. the lazy-bounds calculation based on 'da.stack' could be done more neatly by using flattened indices and then reshaping
 
I have since corrected (1.) and tried (2.) as an alternative calculation method.

This enabled me to prove that partial fetches of mesh_coord points and bounds values **_can_** result in  fetching only _part_ of the source data arrays. 
I also introduced some non-trivial chunking, and investigated the effects of that too.
The results are now documented in the code, with some alternatives and their effects noted.

So I've fixed the tests to specific results, and all pass.
But I have left in the code for some unused alternatives,  which we would not wish to merge as-is.
So I think this Draft PR now documents the investigation + some findings, but is obviously not suitable for merging.

If we want to merge some of this code as an actual solution, we should cut this down + present in a new PR.

pp-mo commented:
More ideas here : https://github.com/pp-mo/iris/pull/65
This is very much a performance-driven implementation detail.  But I think it show the nasty problem that Dask has here, and one  way around it.

pp-mo commented:
Closing, as this PR has served its purpose as a basis for discussion.
We will include some of these ideas in forthcoming development to add a MeshCoord class.
